l1-contracts/contracts/upgrades/GatewayUpgrade.sol

```diff
@@ -0,0 +1,43 @@
+// SPDX-License-Identifier: MIT
+
+pragma solidity 0.8.24;
+
+import {Diamond} from "../state-transition/libraries/Diamond.sol";
+import {BaseZkSyncUpgrade, ProposedUpgrade} from "./BaseZkSyncUpgrade.sol";
+
+import {DataEncoding} from "../common/libraries/DataEncoding.sol";
+
+import {AdminFacet} from "../state-transition/chain-deps/facets/Admin.sol";
+
+/// @author Matter Labs
+/// @custom:security-contact security@matterlabs.dev
+/// @notice This upgrade will be used to migrate Era to be part of the hyperchain ecosystem contracts.
+contract GatewayUpgrade is BaseZkSyncUpgrade, AdminFacet {
+    /// @notice The main function that will be called by the upgrade proxy.
+    /// @param _proposedUpgrade The upgrade to be executed.
+    function upgrade(ProposedUpgrade calldata _proposedUpgrade) public override returns (bytes32) {
+        (
+            address l1DAValidator,
+            address l2DAValidator,
+            address l2LegacyBridge,
+            bytes memory l2TxDataStart,
+            bytes memory l2TxDataFinish
+        ) = abi.decode(_proposedUpgrade.postUpgradeCalldata, (address, address, address, bytes, bytes));
+
+        s.baseTokenAssetId = DataEncoding.encodeNTVAssetId(block.chainid, s.baseToken);
+        /// maybe set baseTokenAssetId in Bridgehub here
+        _setDAValidatorPair(l1DAValidator, l2DAValidator);
+
+        ProposedUpgrade memory proposedUpgrade = _proposedUpgrade;
+        /// We might want to read the l2LegacyBridge address from a specific contract based on chainId.
+        proposedUpgrade.l2ProtocolUpgradeTx.data = bytes.concat(l2TxDataStart, bytes20(l2LegacyBridge), l2TxDataFinish);
```

Comment:
I have a feeling that working with `bytes32(l2LegacyBridge)` will be easier later on, since it is a part of constructor and so we could abi-encode separate elements and then concatenate those, though we can always change it later